lib/cretonne/src/ssa.rs

```diff
@@ -0,0 +1,755 @@
+///
+/// The parameter struct `Variable` corresponds to the way variables are represented in the
+/// non-SSA language you're translating from.
+///
+/// The SSA building relies on information about the variables used and defined, as well as
+/// their position relative to basic blocks which are stricter than extended basic blocks since
+/// they don't allow branching in the middle of them.
+///
+/// This SSA building module allows you to def and use variables on the fly while you are
+/// constructing the CFG, no need for a separate SSA pass after the CFG is completed.
+///
+/// A basic block is said _filled_ if all the instruction that it contains have been translated,
+/// and it is said _sealed_ if all of its predecessors have been declared. Only filled predecessors
+/// can be declared.
+pub struct SSABuilder<Variable>
+    where Variable: EntityRef + Default
+{
+    // Records for every variable and for every revelant block, the last definition of
+    // the variable in the block.
+    variables: EntityMap<Variable, HashMap<Block, Value>>,
+    // Records the position of the basic blocks and the list of values used but not defined in the
+    // block.
+    blocks: EntityMap<Block, BlockData<Variable>>,
+    // Records the basic blocks at the beginning of the `Ebb`s.
+    // The value is `(Ebb,Block)` because in a `SparseMap`, the key has to be computable from the
+    // value.
+    ebb_headers: EntityMap<Ebb, PackedOption<Block>>,
+}
+
+// Describes the current position of a basic block in the control flow graph.
+enum BlockData<Variable> {
+    // A block at the top of an `Ebb`.
+    EbbHeader(EbbHeaderBlockData<Variable>),
+    // A block inside an `Ebb` with an unique other block as its predecessor.
+    // The block is implicitely sealed at creation.
+    // The fields are `(predecessor, ebb)`.
+    EbbBody(Block, Ebb),
+}
+impl<Variable> PrimaryEntityData for BlockData<Variable> {}
+
+impl<Variable> BlockData<Variable> {
+    fn add_predecessor(&mut self, pred: Block, inst: Inst) {
+        match self {
+            &mut BlockData::EbbBody(_, _) => assert!(false),
+            &mut BlockData::EbbHeader(ref mut data) => {
+                data.predecessors.insert(pred, inst);
+                ()
+            }
+        }
+    }
+    fn remove_predecessor(&mut self, inst: Inst) -> Block {
+        match self {
+            &mut BlockData::EbbBody(_, _) => panic!("should not happen"),
+            &mut BlockData::EbbHeader(ref mut data) => {
+                // This a linear complexity operation but the number of predecessors is low
+                // in all non-pathological cases
+                let pred: Block = match data.predecessors
+                          .iter()
+                          .find(|&(_, &jump_inst)| jump_inst == inst) {
+                    None => panic!("the predecessor you are trying to remove is not declared"),
+                    Some((&b, _)) => b.clone(),
+                };
```

Comment:
Actually no because it is the basic block we need here and not the extended basic block.